src/dotnet/CarbonAware.Tools.WattTimeClient/WattTimeClient.cs

```diff
@@ -0,0 +1,327 @@
+
+        return JsonSerializer.Deserialize<List<Forecast>>(result, options) ?? new List<Forecast>();
+    }
+
+    /// <inheritdoc/>
+    public Task<IEnumerable<Forecast>> GetForecastByDateAsync(BalancingAuthority balancingAuthority, string startTime, string endTime)
+    {
+        return this.GetForecastByDateAsync(balancingAuthority.Abbreviation, startTime, endTime);
+    }
+
+    /// <inheritdoc/>
+    public async Task<BalancingAuthority?> GetBalancingAuthorityAsync(string latitude, string longitude)
+    {
+        Log.LogInformation("Requesting balancing authority for lattitude {lattitude} and longitude {longitude}", latitude, longitude);
+
+        var parameters = new Dictionary<string, string>()
+        {
+            { QueryStrings.Latitude, latitude },
+            { QueryStrings.Longitude, longitude }
+        };
+
+        var tags = new Dictionary<string, string>()
+        {
+            { QueryStrings.Latitude, latitude },
+            { QueryStrings.Longitude, longitude }
+        };
+
+        var result = await this.MakeRequestAsync(Paths.BalancingAuthorityFromLocation, parameters, tags);
+
+        return JsonSerializer.Deserialize<BalancingAuthority>(result, options);
+    }
+
+    /// <inheritdoc/>
+    public async Task<string?> GetBalancingAuthorityAbbreviationAsync(string latitude, string longitude)
+    {
+        return (await this.GetBalancingAuthorityAsync(latitude, longitude))?.Abbreviation;
+    }
+
+    /// <inheritdoc/>
+    public async Task<Stream> GetHistoricalDataAsync(string balancingAuthorityAbbreviation)
+    {
+        Log.LogInformation("Requesting historical data for balancing authority {balancingAuthority}", balancingAuthorityAbbreviation);
+
+        var parameters = new Dictionary<string, string>()
+        {
+            { QueryStrings.BalancingAuthorityAbbreviation, balancingAuthorityAbbreviation }
+        };
+
+        using (var activity = ActivitySource.StartActivity(nameof(WattTimeClient)))
+        {
+            var url = BuildUrlWithQueryString(Paths.Historical, parameters);
+
+            Log.LogInformation("Requesting data using url {url}", url);
+            activity?.AddTag(QueryStrings.BalancingAuthorityAbbreviation, balancingAuthorityAbbreviation);
+
+            var result = await this.GetAsyncStreamWithAuthRetry(url);
+
+            Log.LogDebug("For query {url}, received data stream", url);
+
+            return result;
+        }
+    }
+
+    /// <inheritdoc/>
+    public Task<Stream> GetHistoricalDataAsync(BalancingAuthority balancingAuthority)
+    {
+        return this.GetHistoricalDataAsync(balancingAuthority.Abbreviation);
+    }
+
+    private async Task<HttpResponseMessage> GetAsyncWithAuthRetry(string uriPath)
+    {
+        await this.EnsureTokenAsync();
+
+        var response = await this.client.GetAsync(uriPath);
+
+        if (!response.IsSuccessStatusCode && RetriableStatusCodes.Contains(response.StatusCode))
```

Comment:
Checking for success is not needed here. Only for the expected status code. This would simplify the `if`

Reply:
Fixed.